test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java

```diff
@@ -50,13 +54,96 @@ public class AbstractVectorLoadStoreTest extends AbstractVectorTest {
     );
 
     static final List<IntFunction<MemorySegment>> MEMORY_SEGMENT_GENERATORS = List.of(
-            withToString("HMS", (int s) ->
+            withToString("DMS", (int s) ->
                     Arena.ofAuto().allocate(s)
             ),
-            withToString("DMS", (int s) -> {
+            withToString("HMS:byte[]", (int s) -> {
                 byte[] b = new byte[s];
                 return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:short[]", (int s) -> {
+                short[] b = new short[s / Short.BYTES];
+                return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:int[]", (int s) -> {
+                int[] b = new int[s / Integer.BYTES];
+                return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:float[]", (int s) -> {
+                float[] b = new float[s / Float.BYTES];
+                return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:long[]", (int s) -> {
+                long[] b = new long[s / Long.BYTES];
+                return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:double[]", (int s) -> {
+                double[] b = new double[s / Double.BYTES];
+                return MemorySegment.ofArray(b);
+            }),
+            withToString("HMS:ByteBuffer.wrap", (int s) -> {
+                byte[] b = new byte[s];
+                ByteBuffer buff = ByteBuffer.wrap(b);
+                return MemorySegment.ofBuffer(buff);
+            }),
+            // Just test one of the specialized buffers
+            withToString("HMS:IntBuffer.wrap", (int s) -> {
+                int[] b = new int[s / Integer.BYTES];
+                IntBuffer buff = IntBuffer.wrap(b);
+                return MemorySegment.ofBuffer(buff);
+            }),
+            withToString("HMS:ByteBuffer.allocate", (int s) -> {
+                ByteBuffer buff = ByteBuffer.allocate(s);
+                return MemorySegment.ofBuffer(buff);
+            }),
+            // Just test one of the specialized buffers
+            withToString("HMS:IntBuffer.allocate", (int s) -> {
+                IntBuffer buff = IntBuffer.allocate(s / Integer.BYTES);
+                return MemorySegment.ofBuffer(buff);
+            }),
+            // Slice
+            withToString("HMS:long[].asSlice", (int s) -> {
+                long[] b = new long[s / Long.BYTES + 1];
+                return MemorySegment.ofArray(b).asSlice(Long.BYTES);
             })
     );
 
+    static Stream<IntFunction<MemorySegment>> memorySegmentGenerators(ValueLayout elementLayout) {
+        return MEMORY_SEGMENT_GENERATORS.stream()
+                .filter(f -> canBeConverted(f, elementLayout));
+    }
+
+    private static boolean canBeConverted(IntFunction<MemorySegment> function, ValueLayout elementLayout) {
+        // Create a sample to analyze
+        MemorySegment s = function.apply(Long.BYTES);
```

Comment:
I believe that a good way to test this is the following:
* each vector type operates as having a given element layout - for instance, you can imagine the layout for IntVector to be `JAVA_INT` and so forth
* asking whether you load a segment into a vector is the same as asking whether you can access the segment, at offset 0L, with the layout associated with the vector (see above) - that is if MS::get throws, then vector load should also throw (and viceversa)

Reply:
Vectors use and have in their docs, a layout with a byte alignment of 1. e.g. from `IntVector`

```
 ValueLayout.OfInt ELEMENT_LAYOUT = ValueLayout.JAVA_INT.withByteAlignment(1)
```

Can the `fromMemorySegment` not just behave similarly? I get that alignment is preferable, but does it need to be enforced?  If so, then maybe the `ELEMENT_LAYOUT` and example in the javadoc needs to be updated.

Reply:
I've removed the alignment checks.